developing/cicd/README.md

```diff
@@ -8,9 +8,175 @@ actually be used by anyone. This section provides resources for setting
+
+**Deployment should be repeatable.** The same code and configuration should be deployable again and again.
+
+**Deployment process for software should be the same regardless of environment.** Using the same process and tooling means lower cognitive overhead in deployment and your non production environments should reflect production more closely.
+
+**Deployment tools should be maintained separate from the application being deployed.** We've seen this happen where we need to fix some deployed code but also had a problem with the code that performs the deployment and all of it is a hairy mess of debugging.
+
+### Release
+
+There are a few different ways of implementing release.
+No matter what, release is the step when you allow customers to have access to that particular version of your code.
+
+If you're releasing a CLI, it may mean that you're releasing your binary to a Homebrew tap or just a zip file to GitHub releases.
+
+If you have a service, you could have a loadbalancer manage traffic between different versions of your services.
+Or you could use something like Launch Darkly to leverage feature flags to do something similar for different user cohorts.
+
+## Continous (Integration | Delivery | Deploy)
+
+We throw around terms like CI or CD with the assumption that we know what the differences are between these practices.
+
+For clarity, at Truss we'll refer to them as follows:
+
+### Continuous Integration
+
+`Continuous Integration` is when developers merge changes into mainline as often as possible.
+Automated build and test steps validate changes and are incorporated into review before merging to mainline.
+
+This helps mitigate risks from long lived branches but you have automated tests and a developer workflow that allows for frequent merges.
+
+Small incremental changes to code and small discrete tests facilitate freqent merges in CI.
+
+### Continuous Delivery
+
+`Continuous Delivery` can be described as an extension to `Continuous Integration`.
+In addition to automated build and test steps running on every proposed change to mainline, the delivery of the code is also automated.
+
+You want the software built to be deployable at any given point in time.
+
+Delivery may mean different things depending on your workflow.
+It may mean that you have built your artifacts and validated them and delivered them to a centralized repository.
+For example, Docker containers might go to DockerHUB or ECR.
+Might mean that the code has been deployed to a staging environment.
+All of these could be "delivered" depending on the project's workflow.
+
+### Continous Deployment
+
+`Continuous Delivery` can be described as a further extension of `Continuous Delivery`.
+In addition to automated build, test, and delivery production deployment is also automated.
+
+Generally speaking, you will want to have some good logging and monitoring instrumentation so that you can automate rolling back or forward when something goes wrong.
```

Comment:
I'd want to include something like …

This is an advanced state that requires excellent, trustworthy, automated tests and monitoring. The complexity of continuous deployment system is correlated with the complexity of the delivery pipeline. It's much easier to have continous deployment of a single container image than a full-feature web application.